site/_posts/2020-08-19-theme-registration-form.md

```diff
@@ -0,0 +1,285 @@
+
+## Adding form labels
+
+You can customize your field display more extensively by modifying the `customField` macro. This is in the "Helpers" section of your theme. Rather than use the API to do this, let's edit it directly in the administrative user interface.
+
+Navigate to "Themes" and edit your theme. Click on "Helpers" and scroll to the bottom. You'll be modifying the `customField` macro. Here's the default implementation for this freemarker macro:
+
+```
+[#macro customField field key autofocus=false placeholder=""]
+  [#assign fieldId = field.key?replace(".", "_") /]
+  [#local leftAddon = field.data.leftAddon!'info' /]
+  [#if field.control == "checkbox"]
+    [#if field.options?has_content]
+      [@checkbox_list field=field id="${fieldId}" name="${key}" required=field.required autofocus=autofocus options=field.options /]
+    [#else]
+      [@checkbox field=field id="${fieldId}" name="${key}" required=field.required autofocus=autofocus /]
+    [/#if]
+  [#elseif field.control == "number"]
+    [@input id="${fieldId}" type="number" name="${key}" leftAddon="${leftAddon}" required=field.required autofocus=autofocus placeholder=theme.optionalMessage(placeholder) /]
+  [#elseif field.control == "password"]
+    [@input id="${fieldId}" type="password" name="${key}" leftAddon="lock" autocomplete="new-password" autofocus=autofocus placeholder=theme.optionalMessage(placeholder)/]
+  [#elseif field.control == "radio"]
+    [@radio_list field=field id="${fieldId}" name="${key}" required=field.required autofocus=autofocus options=field.options /]
+  [#elseif field.control == "select"]
+    [@select id="${fieldId}" name="${key}" required=field.required autofocus=autofocus options=field.options /]
+  [#elseif field.control == "textarea"]
+    [@textarea id="${fieldId}" name="${key}" required=field.required autofocus=autofocus placeholder=theme.optionalMessage(placeholder) /]
+  [#elseif field.control == "text"]
+    [#if field.type == "date"]
+      [@input id="${fieldId}" type="text" name="${key}" leftAddon="${leftAddon}" required=field.required autofocus=autofocus placeholder=theme.optionalMessage(placeholder) class="date-picker" dateTimeFormat="yyyy-MM-dd" /]
+    [#else]
+      [@input id="${fieldId}" type="text" name="${key}" leftAddon="${leftAddon}" required=field.required autofocus=autofocus placeholder=theme.optionalMessage(placeholder) /]
+    [/#if]
+  [/#if]
+[/#macro]
+```
+
+It looks a little wonky if you aren't used to it, but this is a series of if/then statements which get executed against every custom field as the user interface is being generated. The macro looks at each field definition and creates the correct HTML element. Well, technically, it calls another macro such as `input` or `select`, but you get what I mean. For instance, a `password` field will be rendered as an HTML input field with the type `password`. 
+
+Let's add a label to each field. Right after `[#assign fieldId = field.key?replace(".", "_") /]`, add this:
```

Comment:
This section works, but do you want to make these localizable? If you do, they they should actually use the ${fieldId} to get a message key. This will use all of the messages that you created above. The code would look like this:

```
<label for="${fieldId}">${theme.optionalMessage(fieldId)}:</label>
```

The benefit to doing it this way is there are also no more if-then statements. Just a single line of code that works for everything.

Reply:
Sick! That's great.